MPU/src/driverio.cpp

```diff
@@ -6,102 +6,76 @@ DRIVERIO::DRIVERIO(){}
 
     powerToggle_wait.cancelTimer();
-    ssButton_debounce.cancelTimer();
-    speaker_wait.cancelTimer();
-    tempWarningBlink_wait.cancelTimer();
 
     motorController = p_motorController;
     bms = p_bms;
+
+    //Initializes the motor direction based on the initial switch state
+    if(reverseSwitch.getSwitchState()) motorController->toggleDirection();
 }
 
 
 DRIVERIO::~DRIVERIO(){}
 
 
-void DRIVERIO::handleSSButton()
+void DRIVERIO::handleSSButtonPress()
 {
-    //if the button is still being held during and after the timer runs out, then toggle power
-    if(!ssButton_debounce.isTimerExpired())
-    {
-        if(!digitalRead(SS_BUTT_PIN)) // If released
-        {
-            ssButton_debounce.cancelTimer();
-        }
-        return;
-    }
-    if(ssButtonDebounced() && bms->getChargeMode())
+    //Poll Button
+    ssButton.checkButtonPin();
+
+    //Poll Speaker if it is written high
+    speaker.attemptToStopSpeaker();
+
+    //If the button isn't pressed, do nothing
+    if(!ssButton.isButtonPressed_Pulse()) return;
+
+    //If the BMS is charging
+    if(bms->getChargeMode())
     {
         bms->toggleAIR();
         powerToggle_wait.startTimer(1500);
-        ssButton_debounce.cancelTimer();
         return;
     }
-    if(ssButtonDebounced() && (motorController->getIsOn() || (!motorController->getIsOn() && !motorController->checkFault())))
+
+    //If the motor controller is on OR the motor is off and it is ready to go
+    if(motorController->getIsOn() || 
+        (!motorController->getIsOn() && !motorController->checkFault()))
     {
         motorController->togglePower();    //Writes the power state of the motor to the MC message to be sent
         if(motorController->getIsOn())
         {
-            writeSpeaker(HIGH);
-            speaker_wait.startTimer(1500);
+            speaker.playSpeaker();
         }
         powerToggle_wait.startTimer(1500);
-        ssButton_debounce.cancelTimer();
         return;
     }
-    if(speaker_wait.isTimerExpired())
-    {
-        writeSpeaker(LOW);
-    }
-    if(digitalRead(SS_BUTT_PIN) && powerToggle_wait.isTimerExpired()) // If pressed and no timer
-    {
-        ssButton_debounce.startTimer(50);
-    }
-}
-
-
-bool DRIVERIO::ssButtonDebounced()
-{
-    return (digitalRead(SS_BUTT_PIN) && ssButton_debounce.isTimerExpired() && !ssButton_debounce.isTimerReset());
 }
 
 
 void DRIVERIO::handleSSLED()
 {
     if(bms->getChargeMode())
     {
-        digitalWrite(SS_LED_PIN,!(bms->isAIROpen()));
+        ssButton.writeLED(!bms->isAIROpen());
     }
     else
     {
-        digitalWrite(SS_LED_PIN,motorController->getIsOn());
+        ssButton.writeLED(motorController->getIsOn());
     }
 }
 
 
 void DRIVERIO::handleReverseSwitch()
 {
-    //Serial.print("Switch State:\t");
-    //Serial.println(motorController->getDirection());
-    if(digitalRead(REVERSE_SW_PIN) != motorController->getDirection())
+    if(reverseSwitch.hasSwitchToggled())
```

Comment:
Careful with this. You are changing the way this works to no longer be "ridged". Theoretically if the teensy somehow fucked up and flipped a bit this would make forwards backward and backwards forwards, since you are now just triggering on toggles, rather than literally mapping forward = high, reverse = low

Reply:
99% of the time this should be fine, but I think it may be more robust to continue ridged mapping

Reply:
I changed it to ridged mapping